crates/solvers/src/domain/dex/mod.rs

```diff
@@ -101,20 +104,82 @@ impl Swap {
 
     /// Constructs a single order `solution::Solution` for this swap. Returns
     /// `None` if the swap is not valid for the specified order.
-    pub fn into_solution(self, order: order::Order) -> Option<solution::Solution> {
-        if !self.matches_order(&order) || !self.respects_price(&order) {
+    pub fn into_solution(
+        self,
+        order: order::Order,
+        sell_price: auction::Price,
+        gas: auction::GasPrice,
+    ) -> Option<solution::Solution> {
+        if (order.sell.token, order.buy.token) != (self.input.token, self.output.token) {
             return None;
         }
 
+        let fee = if order.has_solver_fee() {
+            // TODO: If the order has signed `fee` amount already, we should
+            // discount it from the surplus fee. ATM, users would pay both a
+            // full order fee as well as a solver computed fee. Note that this
+            // is fine for now, since there is no way to create limit orders
+            // with non-zero fees.
+            Some(
+                sell_price.ether_value(eth::Ether(
+                    self.gas
+                        .0
+                        .checked_add(Self::SETTLEMENT_OVERHEAD.into())?
+                        .checked_mul(gas.0 .0)?,
+                ))?,
+            )
+        } else {
+            None
+        };
+
+        // Compute total executed sell and buy amounts accounting for solver
+        // fees. That is, the total amount of sell tokens transferred into the
+        // contract and the total buy tokens transferred out of the contract.
+        let (sell, buy) = match order.side {
+            order::Side::Buy => (
+                self.input.amount.checked_add(fee.unwrap_or_default())?,
+                self.output.amount,
+            ),
+            order::Side::Sell => {
+                // We want to collect fees in the sell token, so we need to sell
+                // `fee` more than the DEX swap. However, we don't allow
+                // transferring more than `order.sell.amount` (guaranteed by the
+                // Smart Contract), so we need to cap our executed amount to the
+                // order's limit sell amount and compute the executed buy amount
+                // accordingly.
+                let sell = self
+                    .input
+                    .amount
+                    .checked_add(fee.unwrap_or_default())?
+                    .min(order.sell.amount);
+                let buy = util::math::div_ceil(
+                    sell.checked_sub(fee.unwrap_or_default())?
+                        .checked_mul(self.output.amount)?,
+                    self.input.amount,
+                )?;
+                (sell, buy)
+            }
+        };
+
```

Comment:
Isn't there a missing check against self.input and self.output here? I'm not that familiar with this code but this struct seems to be an arbitrary on chain swap and we're checking whether it can be used to settle this order. In that case I'd expect a check to make sure that the sell amount from the order (without fees that the contract keeps) is at least the swap's input amount. But this check doesn't seem to exist.

Reply:
Executed `sell` and `buy` amounts are computed using `self.input` and `self.output`. We then do a `order.sell * buy >= order.buy * sell` cross-product check to verify that the swap amounts are verified.

Reply:
I added an additional E2E test that verifies that an order won't get settled if the input amounts don't satisfy the order's limit price after applying surplus fees.

Reply:
For me you don't need to add a test. I'm just confused about the semantics of Swap at this point. Doesn't block the PR.

Reply:
419e957cc4f65ee536e45e39e0fa43bfb5e99e4c

Reply:
> but this struct seems to be an arbitrary on chain swap and we're checking whether it can be used to settle this order.

This is exactly it. It produces a `solution::Solution`, which, inside, contains a `solution::trade::Fulfillment` (i.e. a traded order). In order to create one of these, you need:
- An order
- An executed amount
- A fee strategy (protocol or solver-computed surplus fee).

Since `solution::trade::Fulfillment::new` can fail if any of these are invalid, we implicitly check that the swap amount is correct (since we use the swapped amounts to compute the order's executed amount).